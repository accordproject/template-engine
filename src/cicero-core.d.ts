declare module '@accordproject/cicero-core';