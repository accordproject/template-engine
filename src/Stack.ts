/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
interface IStack<T> {
    push(item: T): void;
    pop(): T | undefined;
    peek(): T | undefined;
    size(): number;
  }

class Stack<T> implements IStack<T> {
    private storage: T[] = [];

    constructor(private capacity: number = Infinity) {}

    push(item: T): void {
        if (this.size() === this.capacity) {
            throw Error('Stack has reached max capacity, you cannot add more items');
        }
        this.storage.push(item);
    }

    pop(): T | undefined {
        return this.storage.pop();
    }

    peek(): T | undefined {
        return this.storage[this.size() - 1];
    }

    size(): number {
        return this.storage.length;
    }
}