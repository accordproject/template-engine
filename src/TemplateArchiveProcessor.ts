/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import { Template } from '@accordproject/cicero-core';
import { TemplateMarkInterpreter } from './TemplateMarkInterpreter';
import { TemplateMarkTransformer } from '@accordproject/markdown-template';
import { transform } from '@accordproject/markdown-transform';
import { TypeScriptToJavaScriptCompiler } from './TypeScriptToJavaScriptCompiler';
import Script from '@accordproject/cicero-core/types/src/script';
import { TwoSlashReturn } from '@typescript/twoslash';
import { JavaScriptEvaluator } from './JavaScriptEvaluator';
import { SMART_LEGAL_CONTRACT_BASE64 } from './runtime/declarations';

/**
 * A template archive processor: can draft content using the
 * templatemark for the archive and trigger the logic of the archive
 */
export class TemplateArchiveProcessor {
    template: Template;

    /**
     * Creates a template archive processor
     * @param {Template} template - the template to be used by the processor
     */
    constructor(template: Template) {
        this.template = template;
    }

    /**
     * Drafts a template by merging it with data
     * @param {any} data the data to merge with the template
     * @param {string} format the output format
     * @param {any} options merge options
     * @param {[string]} currentTime the current value for 'now'
     * @returns {Promise} the drafted content
     */
    async draft(data: any, format: string, options: any, currentTime?: string): Promise<any> {
        // Setup
        const metadata = this.template.getMetadata();
        const templateKind = metadata.getTemplateType() !== 0 ? 'clause' : 'contract';

        // Get the data
        const modelManager = this.template.getModelManager();
        const engine = new TemplateMarkInterpreter(modelManager, {});
        const templateMarkTransformer = new TemplateMarkTransformer();
        const templateMarkDom = templateMarkTransformer.fromMarkdownTemplate(
            { content: this.template.getTemplate() }, modelManager, templateKind, {options});
        const now = currentTime ? currentTime : new Date().toISOString();
        // console.log(JSON.stringify(templateMarkDom, null, 2));
        const ciceroMark = await engine.generate(templateMarkDom, data, { now });
        // console.log(JSON.stringify(ciceroMark));
        const result = transform(ciceroMark.toJSON(), 'ciceromark', ['ciceromark_unquoted', format], null, options);
        // console.log(result);
        return result;

    }

    /**
     * Trigger the logic of a template
     * @param {object} request - the request to send to the template logic
     * @param {object} state - the current state of the template
     * @param {[string]} currentTime - the current time, defaults to now
     * @param {[number]} utcOffset - the UTC offer, defaults to zero
     * @returns {Promise} the response and any events
     */
    async trigger(data: any, request: any, state?: any, currentTime?: string, utcOffset?: number): Promise<any> {
        const logicManager = this.template.getLogicManager();
        if(logicManager.getLanguage() === 'typescript') {
            const compiledCode:Record<string, TwoSlashReturn> = {};
            const tsFiles:Array<Script> = logicManager.getScriptManager().getScriptsForTarget('typescript');
            for(let n=0; n < tsFiles.length; n++) {
                const tsFile = tsFiles[n];
                // console.log(`Compiling ${tsFile.getIdentifier()}`);

                const compiler = new TypeScriptToJavaScriptCompiler(this.template.getModelManager(),
                    this.template.getTemplateModel().getFullyQualifiedName());

                await compiler.initialize();

                // add the runtime type definitions to all ts files??
                const code = `${Buffer.from(SMART_LEGAL_CONTRACT_BASE64, 'base64').toString()}
                ${tsFile.getContents()}`

                const result = compiler.compile(code);
                compiledCode[tsFile.getIdentifier()] = result;
            }
            // console.log(compiledCode['logic/logic.ts'].code);
            const evaluator = new JavaScriptEvaluator();
            const evalResponse = await evaluator.evalDangerously( {
                templateLogic: true,
                verbose: false,
                code: compiledCode['logic/logic.ts'].code, // TODO DCS - how to find the code to run?
                argumentNames: ['data', 'request', 'state'],
                arguments: [data, request, state, currentTime, utcOffset]
            });
            return evalResponse.result;
        }
        else {
            return -1;
        }
    }
}
