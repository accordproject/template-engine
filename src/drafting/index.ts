/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import booleanDrafter from './Boolean';
import dateTimeDrafter from './DateTime';
import doubleDrafter from './Double';
import integerDrafter from './Integer';
import longDrafter from './Long';
import monetaryAmountDrafter from './MonetaryAmount';
import stringDrafter from './String';

export function getDrafter(typeName: string) : ((value:any, format:string) => string)|null  {
    switch(typeName) {
    case 'Boolean': return booleanDrafter;
    case 'DateTime': return dateTimeDrafter;
    case 'Double': return doubleDrafter;
    case 'Integer': return integerDrafter;
    case 'Long': return longDrafter;
    case 'org.accordproject.money@0.3.0.MonetaryAmount': return monetaryAmountDrafter;
    case 'String': return stringDrafter;
    default: return null;
    }
}