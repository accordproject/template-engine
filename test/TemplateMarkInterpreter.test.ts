import { ModelManager } from '@accordproject/concerto-core';
import { CommonMarkModel } from '@accordproject/markdown-common';
import { TemplateMarkInterpreter } from '../src';
import { TemplateMarkTransformer } from '@accordproject/markdown-template';
import { readFileSync, readdirSync } from 'fs';
import * as path from 'path';

const CLAUSE_LIBRARY = {
    'clauses': [
        {
            'category': 'reference',
            'author': 'Dan',
            'body': 'Dad jokes',
            'risk': 8.95
        },
        {
            'category': 'reference',
            'author': 'Dan',
            'body': 'Programming for fun and profit',
            'risk': 1.95
        },
        {
            'category': 'onboarding',
            'author': 'Peter',
            'body': 'Legal stuff',
            'risk': 2.95
        },
        {
            'category': 'offboarding',
            'author': 'Matt',
            'body': 'All goodness',
            'risk': 4.95
        }
    ]
};

const GOOD_TEMPLATES_ROOT = './test/templates/good';
const BAD_TEMPLATES_ROOT = './test/templates/bad';

describe('templatemark interpreter', () => {
    jest.setTimeout(30000);
    const goodTemplates: Array<{ name: string, content: string }> = readdirSync(GOOD_TEMPLATES_ROOT).map(dir => {
        return {
            name: dir,
            content: readFileSync(path.join(GOOD_TEMPLATES_ROOT, dir, 'template.md'), 'utf-8')
        };
    });

    goodTemplates.forEach(function (template) {
        if(template.name === 'foreach') return; // currently broken!!

        test(`should generate ${template.name}`, async () => {
            const templateName = path.parse(template.name).name;

            /**
             * Define the data model for the template. The model must have a concept with
             * the @template decorator. The types of properties allow the template to be
             * type-checked.
            */
            const model = readFileSync(`${GOOD_TEMPLATES_ROOT}/${templateName}/model.cto`, 'utf-8');

            /**
             * Load the template, rich-text with variables, conditional sections etc
             */
            const templateMarkup = readFileSync(`${GOOD_TEMPLATES_ROOT}/${templateName}/template.md`, 'utf-8');

            /**
             * Define the data we will merge with the template - an instance of the template model
             */
            const data = JSON.parse(readFileSync(`${GOOD_TEMPLATES_ROOT}/${templateName}/data.json`, 'utf-8'));

            const modelManager = new ModelManager({ strict: true });
            modelManager.addCTOModel(model, undefined, true);
            await modelManager.updateExternalModels();
            const engine = new TemplateMarkInterpreter(modelManager, CLAUSE_LIBRARY);

            const templateMarkTransformer = new TemplateMarkTransformer();

            const templateMarkDom = templateMarkTransformer.fromMarkdownTemplate({ content: templateMarkup }, modelManager, 'contract', { verbose: false });
            // console.log(JSON.stringify(templateMarkDom, null, 2));

            // execute without sandbox
            const now = '2023-03-17T00:00:00.000Z';
            const ciceroMark = await engine.generate(templateMarkDom, data, {now, locale: 'en'});
            expect(ciceroMark.getFullyQualifiedType()).toBe(`${CommonMarkModel.NAMESPACE}.Document`);
            expect(ciceroMark.toJSON()).toMatchSnapshot();

            // check we get the same result when we execute with sandbox
            const ciceroMarkSandbox = await engine.generate(templateMarkDom, data, {now, locale: 'en', childProcessJavaScriptEvaluation: true});
            expect(ciceroMarkSandbox.toJSON()).toStrictEqual(ciceroMark.toJSON());
        });
    });

    const badTemplates: Array<{ name: string, content: string }> = readdirSync(BAD_TEMPLATES_ROOT).map(dir => {
        return {
            name: dir,
            content: readFileSync(path.join(BAD_TEMPLATES_ROOT, dir, 'template.md'), 'utf-8')
        };
    });

    badTemplates.forEach(function (template) {
        test(`should fail to generate ${template.name}`, async () => {
            const templatenName = path.parse(template.name).name;
            const model = readFileSync(`${BAD_TEMPLATES_ROOT}/${templatenName}/model.cto`, 'utf-8');
            const templateMarkup = readFileSync(`${BAD_TEMPLATES_ROOT}/${templatenName}/template.md`, 'utf-8');
            const data = JSON.parse(readFileSync(`${BAD_TEMPLATES_ROOT}/${templatenName}/data.json`, 'utf-8'));
            const modelManager = new ModelManager({ strict: true });
            modelManager.addCTOModel(model);
            const engine = new TemplateMarkInterpreter(modelManager, CLAUSE_LIBRARY);

            const f = async () => {
                const templateMarkTransformer = new TemplateMarkTransformer();
                const templateMarkDom = templateMarkTransformer.fromMarkdownTemplate({ content: templateMarkup }, modelManager, 'contract', { verbose: false });
                const now = '2023-03-17T00:00:00.000Z';
                return engine.generate(templateMarkDom, data, {now});
            };
            await expect(f()).rejects.toMatchSnapshot();
        });
    });
});
