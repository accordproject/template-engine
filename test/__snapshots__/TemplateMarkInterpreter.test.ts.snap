// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templatemark interpreter should create agreementmark from a template and data 1`] = `
Object {
  "$class": "org.accordproject.commonmark@0.5.0.Document",
  "nodes": Array [
    Object {
      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
      "nodes": Array [
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "1",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Welcome!",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Hello ",
            },
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Variable",
              "elementType": "String",
              "name": "firstName",
              "value": "Dan",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": " ",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Tenzin-Isaac-Mia",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": " ",
            },
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Conditional",
              "isTrue": true,
              "name": "lastName",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Mister",
                },
              ],
              "whenFalse": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Dude",
                },
              ],
              "whenTrue": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Mister",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "!",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Middle Names",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.List",
          "delimiter": "period",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Item",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                      "elementType": "String",
                      "name": "this",
                      "value": "Tenzin",
                    },
                  ],
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Item",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                      "elementType": "String",
                      "name": "this",
                      "value": "Isaac",
                    },
                  ],
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Item",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                      "elementType": "String",
                      "name": "this",
                      "value": "Mia",
                    },
                  ],
                },
              ],
            },
          ],
          "start": "1",
          "tight": "true",
          "type": "ordered",
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Address",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.ciceromark@0.6.0.Clause",
          "elementType": "test@1.0.0.Address",
          "name": "address",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.List",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": "Street: ",
                        },
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "street",
                          "value": "1 Main Street",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": "City: ",
                        },
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "city",
                          "value": "Boston",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.commonmark@0.5.0.Text",
                          "text": "ZIP: ",
                        },
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "zip",
                          "value": "12345",
                        },
                      ],
                    },
                  ],
                },
              ],
              "tight": "true",
              "type": "bullet",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Your city is: ",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Emph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Strong",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "city",
                          "value": "Boston",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": ".",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Account Status",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Conditional",
              "isTrue": false,
              "name": "active",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your account has been deactivated.",
                },
              ],
              "whenFalse": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your account has been deactivated.",
                },
              ],
              "whenTrue": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your account is active.",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Thank you for visiting us ",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Strong",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.ciceromark@0.6.0.Formula",
                      "dependencies": Array [],
                      "name": "formula_53113a901ca88208df47bc83374866e8d497d84099c0f88123c918ff1960b17e",
                      "value": "66",
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": " days ago.",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Linebreak",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your last visit was: ",
                },
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.FormattedVariable",
                  "elementType": "DateTime",
                  "format": "MM/DD/YYYY",
                  "name": "lastVisit",
                  "value": "01/10/2023",
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": ".",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Formula",
              "dependencies": Array [],
              "name": "formula_7841bd178366af21ea233d10968eaf538d12fb7a50cbaf9cbe1b51a00cb0f6a8",
              "value": "[123 : Dan Selman][234 : Isaac Selman][456 : Tenzin Selman][789 : Mi-a Selman]",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Orders",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.List",
          "delimiter": "period",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Item",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Strong",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "sku",
                          "value": "WIDGET-2000",
                        },
                      ],
                    },
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": " : ",
                    },
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Emph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "Integer",
                          "name": "amount",
                          "value": "10",
                        },
                      ],
                    },
                  ],
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Item",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Strong",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "sku",
                          "value": "DOODAH-X",
                        },
                      ],
                    },
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": " : ",
                    },
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Emph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "Integer",
                          "name": "amount",
                          "value": "3",
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
          "start": "1",
          "tight": "true",
          "type": "ordered",
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Optional",
              "elementType": "test@1.0.0.LoyaltyStatus",
              "hasSome": false,
              "name": "loyaltyStatus",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "You do not have a loyalty status.",
                },
              ],
              "whenNone": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "You do not have a loyalty status.",
                },
              ],
              "whenSome": Array [],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.ciceromark@0.6.0.Clause",
          "elementType": "test@1.0.0.Preferences",
          "name": "preferences",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Heading",
              "level": "2",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Favorite Colors",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Text",
                  "text": "Your favorite colors are: ",
                },
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.Formula",
                  "dependencies": Array [],
                  "name": "formula_4b8f7e95470eda90057c4648aac4e4c7abb3f93559ed348246b6a15ec1fea473",
                  "value": "RED and PINK",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Image",
                  "destination": "https://www.litmus.com/wp-content/uploads/2021/02/motion-tween-example.gif",
                  "nodes": Array [],
                  "title": "",
                },
              ],
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.List",
              "delimiter": "period",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "this",
                          "value": "RED",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$class": "org.accordproject.commonmark@0.5.0.Item",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
                      "nodes": Array [
                        Object {
                          "$class": "org.accordproject.ciceromark@0.6.0.Variable",
                          "elementType": "String",
                          "name": "this",
                          "value": "PINK",
                        },
                      ],
                    },
                  ],
                },
              ],
              "start": "1",
              "tight": "true",
              "type": "ordered",
            },
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.Conditional",
                  "isTrue": true,
                  "name": "favoriteColors",
                  "nodes": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": "You like pink!",
                    },
                  ],
                  "whenFalse": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": "Why don't you like PINK!",
                    },
                  ],
                  "whenTrue": Array [
                    Object {
                      "$class": "org.accordproject.commonmark@0.5.0.Text",
                      "text": "You like pink!",
                    },
                  ],
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "2",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Dynamic Query of Clauses",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Onboarding Clauses",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.Formula",
                  "dependencies": Array [],
                  "name": "formula_a8a1d7714d95baa82f730e0105d2f2d0a9fc25ee9b6055058cc09667c01c01ab",
                  "value": "[{\\"category\\":\\"onboarding\\",\\"author\\":\\"Peter\\",\\"body\\":\\"Legal stuff\\",\\"risk\\":2.95}]",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Authored by ",
            },
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Variable",
              "elementType": "String",
              "name": "firstName",
              "value": "Dan",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.Formula",
                  "dependencies": Array [],
                  "name": "formula_daca9cb2f5bc16b65f544e6f408c1e3121d50a3251ec4fbe2f27132818acc3d2",
                  "value": "[{\\"category\\":\\"reference\\",\\"author\\":\\"Dan\\",\\"body\\":\\"Dad jokes\\",\\"risk\\":8.95},{\\"category\\":\\"reference\\",\\"author\\":\\"Dan\\",\\"body\\":\\"Programming for fun and profit\\",\\"risk\\":1.95}]",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "High Risk",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.Formula",
                  "dependencies": Array [],
                  "name": "formula_4c874b2977b5eef204e4987efc0a5594c7cce66403e54c17daa1e1c721d755a0",
                  "value": "[{\\"category\\":\\"reference\\",\\"author\\":\\"Dan\\",\\"body\\":\\"Dad jokes\\",\\"risk\\":8.95},{\\"category\\":\\"offboarding\\",\\"author\\":\\"Matt\\",\\"body\\":\\"All goodness\\",\\"risk\\":4.95}]",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Heading",
          "level": "3",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Low Risk and authored by ",
            },
            Object {
              "$class": "org.accordproject.ciceromark@0.6.0.Variable",
              "elementType": "String",
              "name": "firstName",
              "value": "Dan",
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.BlockQuote",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
              "nodes": Array [
                Object {
                  "$class": "org.accordproject.ciceromark@0.6.0.Formula",
                  "dependencies": Array [],
                  "name": "formula_646a6cadec2125e4fd9e4b756aac72bc618b528967f04d325b28a817774441dd",
                  "value": "[{\\"category\\":\\"reference\\",\\"author\\":\\"Dan\\",\\"body\\":\\"Programming for fun and profit\\",\\"risk\\":1.95}]",
                },
              ],
            },
          ],
        },
        Object {
          "$class": "org.accordproject.commonmark@0.5.0.Paragraph",
          "nodes": Array [
            Object {
              "$class": "org.accordproject.commonmark@0.5.0.Text",
              "text": "Done.",
            },
          ],
        },
      ],
    },
  ],
  "xmlns": "http://commonmark.org/xml/1.0",
}
`;
